src/redis-check-aof.c

```diff
@@ -146,156 +259,413 @@ int readArgc(FILE *fp, long *target) {
     return readLong(fp,'*',target);
 }
 
-off_t process(FILE *fp) {
+int processRESP(FILE *fp, char *filename, int *multi) {
     long argc;
-    off_t pos = 0;
-    int i, multi = 0;
     char *str;
 
-    while(1) {
-        if (!multi) pos = ftello(fp);
-        if (!readAnnotations(fp)) break;
-        if (!readArgc(fp, &argc)) break;
-
-        for (i = 0; i < argc; i++) {
-            if (!readString(fp,&str)) break;
-            if (i == 0) {
-                if (strcasecmp(str, "multi") == 0) {
-                    if (multi++) {
-                        ERROR("Unexpected MULTI");
-                        break;
-                    }
-                } else if (strcasecmp(str, "exec") == 0) {
-                    if (--multi) {
-                        ERROR("Unexpected EXEC");
-                        break;
-                    }
+    if (!readArgc(fp, &argc)) return 0;
+
+    for (int i = 0; i < argc; i++) {
+        if (!readString(fp, &str)) return 0;
+        if (i == 0) {
+            if (strcasecmp(str, "multi") == 0) {
+                if ((*multi)++) {
+                    ERROR("Unexpected MULTI in AOF %s", filename);
+                    zfree(str);
+                    return 0;
+                }
+            } else if (strcasecmp(str, "exec") == 0) {
+                if (--(*multi)) {
+                    ERROR("Unexpected EXEC in AOF %s", filename);
+                    zfree(str);
+                    return 0;
                 }
             }
-            zfree(str);
-        }
-
-        /* Stop if the loop did not finish */
-        if (i < argc) {
-            if (str) zfree(str);
-            break;
         }
+        zfree(str);
     }
 
-    if (feof(fp) && multi && strlen(error) == 0) {
-        ERROR("Reached EOF before reading EXEC for MULTI");
-    }
-    if (strlen(error) > 0) {
-        printf("%s\n", error);
-    }
-    return pos;
+    return 1;
 }
 
-int redis_check_aof_main(int argc, char **argv) {
-    char *filename;
-    int fix = 0;
+int processAnnotations(FILE *fp, char *filename, int last_file) {
+    char buf[AOF_ANNOTATION_LINE_MAX_LEN];
 
-    if (argc < 2) {
-        goto invalid_args;
-    } else if (argc == 2) {
-        filename = argv[1];
-    } else if (argc == 3) {
-        if (!strcmp(argv[1],"--fix")) {
-            filename = argv[2];
-            fix = 1;
-        } else {
-            goto invalid_args;
+    epos = ftello(fp);
+    if (fgets(buf, sizeof(buf), fp) == NULL) {
+        printf("Failed to read annotations from AOF %s, aborting...\n", filename);
+        fclose(fp);
+        exit(1);
+    }
+
+    if (to_timestamp && strncmp(buf, "#TS:", 4) == 0) {
+        time_t ts = strtol(buf+4, NULL, 10);
```

Comment:
I see what you mean, we do need to do this technically. However, what I want to say is that AOF, including manifest, is generated by `redis` (such as the timestamp here), and we don't want anyone to change these contents, otherwise he should be responsible for this behavior. I tried looking for `strtol` usage in existing `redis` code and I found almost no error checking anywhere, I don't know if this is intentional or not. Anyway, I added error checking here.

Reply:
In _config.c_ we do such error checking. In other places we use `string2ll()` which has its own error checking. I suggest you look in _config.c_.

Reply:
OK, thanks.